rpc: Fix gui shutdown when waitfor* cmds are called from RPC console

hebasto opened the pull request:
On master (7eed413e72a236b6f1475a198f7063fd24929e23), if the GUI has been started with`-server=1`, `bitcoin-qt` hangs on shutdown during calling any of the `waitfor*` commands in the GUI RPC console.

This PR suggests minimal changes to fix this bug.

Fix #17495

hebasto committed:
qt: Fix shutdown when waitfor* cmds are called from RPC console

hebasto commented:
While the Travis-to-GitHub connection is broken, here is Travis's [build of this PR](https://travis-ci.org/github/bitcoin/bitcoin/builds/667841108).

promag reviewed (commented):
Concept ACK. Have you considered doing this in `BitcoinApplication::requestShutdown` or even between `app.exec()` and `app.requestShutdown()` in qt/bitcoin.cpp?

hebasto commented:
@promag 
> Have you considered doing this in `BitcoinApplication::requestShutdown` or even between `app.exec()` and `app.requestShutdown()` in qt/bitcoin.cpp?

Yes, I have. `InterruptRPC(); StopRPC();` could be placed just before the following line
https://github.com/bitcoin/bitcoin/blob/5f9cd62f33fb4d440173b9c376cadf4887e81e9d/src/qt/bitcoin.cpp#L316

What are benefits of such approach while it requires to add `#include <rpc/server.h>` to `qt/bitcoin.cpp`?

Also, please note, that #17659 suggests to streamline the `bitcoin-qt` shutdown routine and leave the only `QApplication::exec()` call.

jonasschnelli commented on src/qt/rpcconsole.cpp:
```diff
@@ -687,7 +687,12 @@ void RPCConsole::setClientModel(ClientModel *model)
-        // Client model is being set to 0, this means shutdown() is about to be called.
+        // Client model is being set to nullptr, this means shutdown() is about to be called.
+        if (gArgs.GetBoolArg("-server", false)) {
+            InterruptRPC();
```
This feels to be the wrong place to call shutdown/init functions. Shouldn't this be called through bitcoin.cpp?

hebasto replied:
[Done](https://github.com/bitcoin/bitcoin/pull/18452#issuecomment-611024400).

jonasschnelli commented:
Concept ACK

hebasto commented:
Updated 29f5e02601730298020827c837db23ac6c7a8010  -> 368530de424fcb73541b089ec555a44abe475a6d ([pr18452.01](https://github.com/hebasto/bitcoin/commits/pr18452.01) -> [pr18452.02](https://github.com/hebasto/bitcoin/commits/pr18452.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18452.01..pr18452.02)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/18452#pullrequestreview-383475251)
> Concept ACK. Have you considered doing this in `BitcoinApplication::requestShutdown` or even between `app.exec()` and `app.requestShutdown()` in qt/bitcoin.cpp?

- addressed @jonasschnelli's [comment](https://github.com/bitcoin/bitcoin/pull/18452#discussion_r405414822)
> This feels to be the wrong place to call shutdown/init functions. Shouldn't this be called through bitcoin.cpp?

promag reviewed (commented):
Code review ACK 368530de424fcb73541b089ec555a44abe475a6d, the approach looks better.

hebasto commented on src/qt/bitcoin.cpp:
```diff
@@ -311,6 +312,11 @@ void BitcoinApplication::requestShutdown()
     // Request node shutdown, which can interrupt long operations, like
     // rescanning a wallet.
     m_node.startShutdown();
+    // Stop RPC for clean shutdown if waitfor* commands are executed.
```
There are other `NodeImpl::startShutdown()` call sites that are irrelevant to the problem:
https://github.com/bitcoin/bitcoin/blob/2b9a4a13324ac3f0adf712d01fb8f2ba914cdc8f/src/qt/splashscreen.cpp#L141

and

https://github.com/bitcoin/bitcoin/blob/2b9a4a13324ac3f0adf712d01fb8f2ba914cdc8f/src/qt/splashscreen.cpp#L223

promag replied:
I did check those and TBH I think it's still fine.

hebasto replied:
> I did check those and TBH I think it's still fine.

To be clear, what exactly is "still fine" in your opinion: pr code or your suggestion to move to the `NodeImpl` ?

hebasto replied:
[Implemented](https://github.com/bitcoin/bitcoin/pull/18452#issuecomment-612135003).

hebasto commented:
Updated 368530de424fcb73541b089ec555a44abe475a6d  -> 59b55ccf9aae2f75e86245434be26e1f1a729659 ([pr18452.02](https://github.com/hebasto/bitcoin/commits/pr18452.02) -> [pr18452.03](https://github.com/hebasto/bitcoin/commits/pr18452.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18452.02..pr18452.03)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/18452#discussion_r406707140)
> I think you could move this to `NodeImpl::startShutdown`?

promag commented:
Tested ACK 59b55ccf9aae2f75e86245434be26e1f1a729659.

luke-jr commented on src/rpc/server.cpp:
```diff
@@ -285,21 +287,28 @@ void StartRPC()
 {
+    // This function could be called twice if the GUI has been started with -server=1.
+    assert(!g_rpc_running);
+    if (g_rpc_stopped) return;
     LogPrint(BCLog::RPC, "Stopping RPC\n");
     deadlineTimers.clear();
     DeleteAuthCookie();
+    g_rpc_stopped = true;
```
Race?

hebasto replied:
https://github.com/bitcoin/bitcoin/blob/59b55ccf9aae2f75e86245434be26e1f1a729659/src/rpc/server.cpp#L24

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18452#issuecomment-620300495).

hebasto commented:
Updated 59b55ccf9aae2f75e86245434be26e1f1a729659  -> 2d7b6bc6c9c49ac95c7aa97d9f753aec97456e0b ([pr18452.03](https://github.com/hebasto/bitcoin/commits/pr18452.03) -> [pr18452.04](https://github.com/hebasto/bitcoin/commits/pr18452.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18452.03..pr18452.04)):

- addressed @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/18452#discussion_r413417576)
> Race?

hebasto commented:
Rebased 2d7b6bc6c9c49ac95c7aa97d9f753aec97456e0b  -> da73f1513a637a9f347b64de66564d6cdb2541f8 ([pr18452.04](https://github.com/hebasto/bitcoin/commits/pr18452.04) -> [pr18452.05](https://github.com/hebasto/bitcoin/commits/pr18452.05)) due to the conflict with #18814.

jonasschnelli commented:
utACK da73f1513a637a9f347b64de66564d6cdb2541f8
